File: proto/control-api-mock/src/endpoint.rs
//! `Endpoint` related methods and entities.

use serde::{Deserialize, Serialize};
use smart_default::SmartDefault;

/// P2P mode of [`WebRtcPublishEndpoint`].
#[derive(Debug, Deserialize, Serialize)]
pub enum P2pMode {
    Always,
    Never,
    IfPossible,
}

/// Publishing policy of the video or audio media type in the
/// [`WebRtcPublishEndpoint`].
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize, SmartDefault)]
pub enum PublishPolicy {
    /// Publish this media type if it possible.
    #[default]
    Optional,

    /// Don't start call if this media type can't be published.
    Required,

    /// Media type __must__ not be published.
    ///
    /// Media server will not try to initialize publishing.
    Disabled,
}

/// Settings for the audio media type of the [`WebRtcPublishEndpoint`].
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct AudioSettings {
    /// Publishing policy of the audio media type in the
    /// [`WebRtcPublishEndpoint`].
    #[serde(default)]
    pub publish_policy: PublishPolicy,
}

/// Settings for the video media type of the [`WebRtcPublishEndpoint`].
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct VideoSettings {
    /// Publishing policy of the video media type in the
    /// [`WebRtcPublishEndpoint`].
    #[serde(default)]
    pub publish_policy: PublishPolicy,
}

/// [Control API]'s `WebRtcPublishEndpoint` representation.
///
/// [Control API]: https://tinyurl.com/yxsqplq7
#[derive(Debug, Deserialize, Serialize)]
pub struct WebRtcPublishEndpoint {
    /// ID of [`WebRtcPublishEndpoint`].
    #[serde(skip_deserializing)]
    pub id: String,

    /// Mode of connection for this [`WebRtcPublishEndpoint`].
    pub p2p: P2pMode,

    /// Option to relay all media through a TURN server forcibly.
    #[serde(default)]
    pub force_relay: bool,

    /// Settings for the audio media type of the [`WebRtcPublishEndpoint`].
    #[serde(default)]
    pub audio_settings: AudioSettings,

    /// Settings for the video media type of the [`WebRtcPublishEndpoint`].
    #[serde(default)]
    pub video_settings: VideoSettings,
}

/// [Control API]'s `WebRtcPlayEndpoint` element representation.
///
/// [Control API]: https://tinyurl.com/yxsqplq7
#[derive(Debug, Deserialize, Serialize)]
pub struct WebRtcPlayEndpoint {
    /// ID of this [`WebRtcPlayEndpoint`].
    #[serde(skip_deserializing)]
    pub id: String,

    /// URI in format `local://{room_id}/{member_id}/{endpoint_id}` pointing to
    /// [`WebRtcPublishEndpoint`] which this [`WebRtcPlayEndpoint`] plays.
    pub src: String,

    /// Option to relay all media through a TURN server forcibly.
    #[serde(default)]
    pub force_relay: bool,
}

/// `Endpoint` element representation.
#[derive(Debug, Deserialize, Serialize)]
#[serde(tag = "kind")]
pub enum Endpoint {
    WebRtcPublishEndpoint(WebRtcPublishEndpoint),
    WebRtcPlayEndpoint(WebRtcPlayEndpoint),
}

File: flutter/test/e2e/api/endpoint.dart
import 'package:json_annotation/json_annotation.dart';

part 'endpoint.g.dart';

/// `Endpoint` element representation.
@JsonSerializable()
class Endpoint {
  Map<String, dynamic> toJson() => {};
  Endpoint();
  factory Endpoint.fromJson(Map<String, dynamic> json) {
    if (json.toString().contains('WebRtcPlayEndpoint')) {
      return WebRtcPlayEndpoint.fromJson(json);
    } else {
      return WebRtcPublishEndpoint.fromJson(json);
    }
  }
}

/// [Control API]'s `WebRtcPlayEndpoint` element representation.
///
/// [Control API]: https://tinyurl.com/yxsqplq7
@JsonSerializable()
class WebRtcPlayEndpoint implements Endpoint {
  /// ID of this [WebRtcPlayEndpoint].
  String id = '';

  /// URI in format `local://{room_id}/{member_id}/{endpoint_id}` pointing to
  /// [WebRtcPublishEndpoint] which this [`WebRtcPlayEndpoint`] plays.
  String src;

  /// Option to relay all media through a TURN server forcibly.
  bool force_relay = false;

  WebRtcPlayEndpoint(this.id, this.src);

  factory WebRtcPlayEndpoint.fromJson(Map<String, dynamic> json) =>
      _$WebRtcPlayEndpointFromJson(json);

  @override
  Map<String, dynamic> toJson() {
    var json = _$WebRtcPlayEndpointToJson(this);
    json.addAll({'kind': 'WebRtcPlayEndpoint'});
    return json;
  }
}

/// [Control API]'s `WebRtcPublishEndpoint` representation.
///
/// [Control API]: https://tinyurl.com/yxsqplq7
@JsonSerializable()
class WebRtcPublishEndpoint implements Endpoint {
  /// ID of [WebRtcPublishEndpoint].
  String id;

  /// Mode of connection for this [WebRtcPublishEndpoint].
  P2pMode p2p;

  /// Option to relay all media through a TURN server forcibly.
  bool force_relay = false;

  /// Settings for the audio media type of the [WebRtcPublishEndpoint].
  AudioSettings audio_settings = AudioSettings(PublishPolicy.Optional);

  /// Settings for the video media type of the [WebRtcPublishEndpoint].
  VideoSettings video_settings = VideoSettings(PublishPolicy.Optional);

  WebRtcPublishEndpoint(this.id, this.p2p);

  factory WebRtcPublishEndpoint.fromJson(Map<String, dynamic> json) =>
      _$WebRtcPublishEndpointFromJson(json);

  @override
  Map<String, dynamic> toJson() {
    var json = _$WebRtcPublishEndpointToJson(this);
    json.addAll({'kind': 'WebRtcPublishEndpoint'});
    return json;
  }
}

/// P2P mode of [WebRtcPublishEndpoint].
enum P2pMode {
  /// Send media data peer-to-peer only without a media server.
  Always,

  /// Always send media data through a media server.
  Never,

  /// Send media data peer-to-peer directly if it's possible, otherwise
  /// through a media server.
  IfPossible,
}

/// Publishing policy of the video or audio media type in the
/// [WebRtcPublishEndpoint].
enum PublishPolicy {
  /// Publish this media type if it possible.
  Optional,

  /// Don't start call if this media type can't be published.
  Required,

  /// Media type __must__ not be published.
  ///
  /// Media server will not try to initialize publishing.
  Disabled,
}

/// Settings for the audio media type of the [WebRtcPublishEndpoint].
@JsonSerializable()
class AudioSettings {
  /// Publishing policy of the audio media type in the
  /// [WebRtcPublishEndpoint].
  PublishPolicy publish_policy = PublishPolicy.Optional;

  AudioSettings(this.publish_policy);

  factory AudioSettings.fromJson(Map<String, dynamic> json) =>
      _$AudioSettingsFromJson(json);
  Map<String, dynamic> toJson() => _$AudioSettingsToJson(this);
}

/// Settings for the video media type of the [WebRtcPublishEndpoint].
@JsonSerializable()
class VideoSettings {
  /// Publishing policy of the video media type in the
  /// [WebRtcPublishEndpoint].
  PublishPolicy publish_policy = PublishPolicy.Optional;

  VideoSettings(this.publish_policy);
  factory VideoSettings.fromJson(Map<String, dynamic> json) =>
      _$VideoSettingsFromJson(json);
  Map<String, dynamic> toJson() => _$VideoSettingsToJson(this);
}
